File: the-simple-webservice/lib/the-simple-webservice-stack.ts
import * as cdk from '@aws-cdk/core';
import lambda = require('@aws-cdk/aws-lambda');
import apigw = require('@aws-cdk/aws-apigateway');
import dynamodb = require('@aws-cdk/aws-dynamodb');

export class TheSimpleWebserviceStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    //DynamoDB Table
    const table = new dynamodb.Table(this, 'Hits', {
      partitionKey: { name: 'path', type: dynamodb.AttributeType.STRING }
    });

     // defines an AWS Lambda resource
     const hello = new lambda.Function(this, 'HelloHandler', {
      runtime: lambda.Runtime.NODEJS_12_X,      // execution environment
      code: lambda.Code.asset('lambda'),  // code loaded from the "lambda" directory
      handler: 'lambda.handler',                // file is "lambda", function is "handler"
      environment: {
        HITS_TABLE_NAME: table.tableName
    }
    });

     // grant the lambda role read/write permissions to our table
     table.grantReadWriteData(hello);

    // defines an API Gateway REST API resource backed by our "hello" function.
    new apigw.LambdaRestApi(this, 'Endpoint', {
      handler: hello
    });
  }
}

File: the-simple-webservice/java/cdk/src/main/java/com/cdkpatterns/TheSimpleWebserviceStack.java
package com.cdkpatterns;

import software.amazon.awscdk.core.CfnOutput;
import software.amazon.awscdk.core.Construct;
import software.amazon.awscdk.core.Stack;
import software.amazon.awscdk.core.StackProps;
import software.amazon.awscdk.services.apigatewayv2.HttpApi;
import software.amazon.awscdk.services.apigatewayv2.integrations.LambdaProxyIntegration;
import software.amazon.awscdk.services.dynamodb.Attribute;
import software.amazon.awscdk.services.dynamodb.AttributeType;
import software.amazon.awscdk.services.dynamodb.Table;
import software.amazon.awscdk.services.dynamodb.TableProps;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.Runtime;

import java.util.Map;

public class TheSimpleWebserviceStack extends Stack {
    public TheSimpleWebserviceStack(final Construct scope, final String id) {
        this(scope, id, null);
    }

    public TheSimpleWebserviceStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        Table dynamoDbTable = createDynamoDBTable();
        Function lambda = createLambda(dynamoDbTable.getTableName());
        HttpApi api = createHttpApi(lambda);

        dynamoDbTable.grantReadWriteData(lambda);

        CfnOutput.Builder.create(this, "ApiUrl")
                .description("HTTP API Url")
                .value(api.getUrl())
                .build();
    }

    private HttpApi createHttpApi(Function dynamoLambda) {
        return HttpApi.Builder.create(this, "Endpoint")
                .defaultIntegration(
                        LambdaProxyIntegration.Builder.create()
                                .handler(dynamoLambda)
                                .build())
                .build();
    }

    private Function createLambda(String tableName) {
        return Function.Builder.create(this, "DynamoLambdaHandler")
                .code(Code.fromAsset("./lambda/target/lambda.zip"))
                .handler("com.cdkpatterns.LambdaHandler::handleRequest")
                .runtime(Runtime.JAVA_11)
                .memorySize(1538)
                .environment(Map.of(
                        "HITS_TABLE_NAME", tableName,
                        "REGION", this.getRegion()))
                .build();
    }

    private Table createDynamoDBTable() {
        return new Table(this, "Hits", TableProps.builder()
                .partitionKey(Attribute.builder()
                        .name("path")
                        .type(AttributeType.STRING)
                        .build())
                .build());
    }
}
